### PR TITLE
Fix "Bad Sad error when the tuple has an array of a record"

### LINKED ISSUE #32266: Node transformer crashes for union of tuple singletons
**Description:**
$title.

**Steps to reproduce:**
```ballerina
public function main() {
    // [1, "hello"] a; // OK
    // [2] b; // OK
    // [1, "hello"]|int d; // OK
    // [string, boolean]|[1] e; // OK

    // Crashes
    [1, "hello"]|[1] f;
}
```

```log
[2021-08-22 11:39:25,090] SEVERE {b7a.log.crash} - Node not supported: BasicLiteralNode 
java.lang.RuntimeException: Node not supported: BasicLiteralNode
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transformSyntaxNode(BLangNodeTransformer.java:3849)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transformSyntaxNode(BLangNodeTransformer.java:473)
        at io.ballerina.compiler.syntax.tree.NodeTransformer.transform(NodeTransformer.java:336)
        at io.ballerina.compiler.syntax.tree.BasicLiteralNode.apply(BasicLiteralNode.java:46)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createTypeNode(BLangNodeTransformer.java:5433)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:997)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:473)
        at io.ballerina.compiler.syntax.tree.TupleTypeDescriptorNode.apply(TupleTypeDescriptorNode.java:54)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createTypeNode(BLangNodeTransformer.java:5433)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:910)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:473)
        at io.ballerina.compiler.syntax.tree.UnionTypeDescriptorNode.apply(UnionTypeDescriptorNode.java:54)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createTypeNode(BLangNodeTransformer.java:5433)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createBLangVarDef(BLangNodeTransformer.java:2807)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:2774)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:473)
        at io.ballerina.compiler.syntax.tree.VariableDeclarationNode.apply(VariableDeclarationNode.java:67)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.generateAndAddBLangStatements(BLangNodeTransformer.java:4948)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:1543)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:473)
        at io.ballerina.compiler.syntax.tree.FunctionBodyBlockNode.apply(FunctionBodyBlockNode.java:59)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.populateFunctionNode(BLangNodeTransformer.java:1450)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createFunctionNode(BLangNodeTransformer.java:1430)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:1408)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:473)
        at io.ballerina.compiler.syntax.tree.FunctionDefinitionNode.apply(FunctionDefinitionNode.java:71)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:607)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:473)
        at io.ballerina.compiler.syntax.tree.ModulePartNode.apply(ModulePartNode.java:54)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.accept(BLangNodeTransformer.java:506)
        at io.ballerina.projects.DocumentContext.compilationUnit(DocumentContext.java:111)
        at io.ballerina.projects.ModuleContext.compileInternal(ModuleContext.java:378)
        at io.ballerina.projects.ModuleCompilationState$1.compile(ModuleCompilationState.java:45)
        at io.ballerina.projects.ModuleContext.compile(ModuleContext.java:341)
        at io.ballerina.projects.PackageCompilation.compileModulesInternal(PackageCompilation.java:199)
        at io.ballerina.projects.PackageCompilation.compileModules(PackageCompilation.java:186)
        at io.ballerina.projects.PackageCompilation.compile(PackageCompilation.java:111)
        at io.ballerina.projects.PackageCompilation.from(PackageCompilation.java:100)
        at io.ballerina.projects.PackageContext.getPackageCompilation(PackageContext.java:214)
        at io.ballerina.projects.Package.getCompilation(Package.java:140)
        at io.ballerina.cli.task.CompileTask.execute(CompileTask.java:75)
        at io.ballerina.cli.TaskExecutor.executeTasks(TaskExecutor.java:40)
        at io.ballerina.cli.cmd.RunCommand.execute(RunCommand.java:176)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at io.ballerina.cli.launcher.Main.main(Main.java:53)
```

**Affected Versions:**
slbeta2, slbeta3-SNAPSHOT

### LINKED ISSUE #32367: Bad Sad error when the tuple has an array of a record
**Description:**
Bad Sad error occurred for the following sample codes:

**Sample 1:**
```ballerina
type Foo record {|
    int id;
|};

type Bar record {|
    string name;
|};

function myfunc() returns Foo|[Foo[], Bar]? {
	
}

public function main() {
    Foo|[Foo[], Bar]? res = myfunc();
}
```

Output:
```shell
ballerina: Oh no, something really went wrong. Bad. Sad.

We appreciate it if you can report the code that broke Ballerina in 
https://github.com/ballerina-platform/ballerina-lang/issues with the 
log you get below and your sample code.

We thank you for helping make us better.

[2021-08-26 09:28:14,135] SEVERE {b7a.log.crash} - class org.wso2.ballerinalang.compiler.tree.expressions.BLangIndexBasedAccess cannot be cast to class org.wso2.ballerinalang.compiler.tree.types.BLangType (org.wso2.ballerinalang.compiler.tree.expressions.BLangIndexBasedAccess and org.wso2.ballerinalang.compiler.tree.types.BLangType are in unnamed module of loader 'app') 
java.lang.ClassCastException: class org.wso2.ballerinalang.compiler.tree.expressions.BLangIndexBasedAccess cannot be cast to class org.wso2.ballerinalang.compiler.tree.types.BLangType (org.wso2.ballerinalang.compiler.tree.expressions.BLangIndexBasedAccess and org.wso2.ballerinalang.compiler.tree.types.BLangType are in unnamed module of loader 'app')
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createTypeNode(BLangNodeTransformer.java:5343)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:961)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:473)
        at io.ballerina.compiler.syntax.tree.TupleTypeDescriptorNode.apply(TupleTypeDescriptorNode.java:54)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createTypeNode(BLangNodeTransformer.java:5343)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:3032)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:473)
        at io.ballerina.compiler.syntax.tree.OptionalTypeDescriptorNode.apply(OptionalTypeDescriptorNode.java:50)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createTypeNode(BLangNodeTransformer.java:5343)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:874)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:473)
        at io.ballerina.compiler.syntax.tree.UnionTypeDescriptorNode.apply(UnionTypeDescriptorNode.java:54)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createTypeNode(BLangNodeTransformer.java:5343)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createBLangVarDef(BLangNodeTransformer.java:2750)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:2717)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:473)
        at io.ballerina.compiler.syntax.tree.VariableDeclarationNode.apply(VariableDeclarationNode.java:67)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.generateAndAddBLangStatements(BLangNodeTransformer.java:4872)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:1507)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:473)
        at io.ballerina.compiler.syntax.tree.FunctionBodyBlockNode.apply(FunctionBodyBlockNode.java:59)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.populateFunctionNode(BLangNodeTransformer.java:1414)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createFunctionNode(BLangNodeTransformer.java:1394)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:1372)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:473)
        at io.ballerina.compiler.syntax.tree.FunctionDefinitionNode.apply(FunctionDefinitionNode.java:71)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:590)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:473)
        at io.ballerina.compiler.syntax.tree.ModulePartNode.apply(ModulePartNode.java:54)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.accept(BLangNodeTransformer.java:503)
        at io.ballerina.projects.DocumentContext.compilationUnit(DocumentContext.java:113)
        at io.ballerina.projects.ModuleContext.compileInternal(ModuleContext.java:365)
        at io.ballerina.projects.ModuleCompilationState$1.compile(ModuleCompilationState.java:45)
        at io.ballerina.projects.ModuleContext.compile(ModuleContext.java:329)
        at io.ballerina.projects.PackageCompilation.compileModulesInternal(PackageCompilation.java:174)
        at io.ballerina.projects.PackageCompilation.compileModules(PackageCompilation.java:166)
        at io.ballerina.projects.PackageCompilation.from(PackageCompilation.java:95)
        at io.ballerina.projects.PackageContext.getPackageCompilation(PackageContext.java:206)
        at io.ballerina.projects.Package.getCompilation(Package.java:135)
        at io.ballerina.cli.task.CompileTask.execute(CompileTask.java:64)
        at io.ballerina.cli.TaskExecutor.executeTasks(TaskExecutor.java:40)
        at io.ballerina.cli.cmd.RunCommand.execute(RunCommand.java:163)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at io.ballerina.cli.launcher.Main.main(Main.java:58)
```

**Sample 2:**
```ballerina
type Foo record {|
    int id;
};

type Bar record {|
    string name;
|};

function myfunc() returns [Foo[], Bar]? {
	
}

public function main() {
    [Foo[], Bar]? res = myfunc();
}
```

Output:
```shell
ballerina: Oh no, something really went wrong. Bad. Sad.

We appreciate it if you can report the code that broke Ballerina in 
https://github.com/ballerina-platform/ballerina-lang/issues with the 
log you get below and your sample code.

We thank you for helping make us better.

[2021-08-26 09:30:12,339] SEVERE {b7a.log.crash} - Index: '0', Size: '0' 
java.lang.IndexOutOfBoundsException: Index: '0', Size: '0'
        at io.ballerina.compiler.internal.syntax.NodeListUtils.rangeCheck(NodeListUtils.java:38)
        at io.ballerina.compiler.syntax.tree.SeparatedNodeList.get(SeparatedNodeList.java:45)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:2156)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:473)
        at io.ballerina.compiler.syntax.tree.IndexedExpressionNode.apply(IndexedExpressionNode.java:58)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createActionOrExpression(BLangNodeTransformer.java:5000)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createExpression(BLangNodeTransformer.java:4971)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:1917)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:473)
        at io.ballerina.compiler.syntax.tree.ListConstructorExpressionNode.apply(ListConstructorExpressionNode.java:54)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createActionOrExpression(BLangNodeTransformer.java:5000)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createExpression(BLangNodeTransformer.java:4971)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:1825)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:473)
        at io.ballerina.compiler.syntax.tree.ConditionalExpressionNode.apply(ConditionalExpressionNode.java:62)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createActionOrExpression(BLangNodeTransformer.java:5000)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createExpression(BLangNodeTransformer.java:4971)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:2836)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:473)
        at io.ballerina.compiler.syntax.tree.ExpressionStatementNode.apply(ExpressionStatementNode.java:50)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.generateAndAddBLangStatements(BLangNodeTransformer.java:4872)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:1507)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:473)
        at io.ballerina.compiler.syntax.tree.FunctionBodyBlockNode.apply(FunctionBodyBlockNode.java:59)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.populateFunctionNode(BLangNodeTransformer.java:1414)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createFunctionNode(BLangNodeTransformer.java:1394)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:1372)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:473)
        at io.ballerina.compiler.syntax.tree.FunctionDefinitionNode.apply(FunctionDefinitionNode.java:71)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:590)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:473)
        at io.ballerina.compiler.syntax.tree.ModulePartNode.apply(ModulePartNode.java:54)
        at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.accept(BLangNodeTransformer.java:503)
        at io.ballerina.projects.DocumentContext.compilationUnit(DocumentContext.java:113)
        at io.ballerina.projects.ModuleContext.compileInternal(ModuleContext.java:365)
        at io.ballerina.projects.ModuleCompilationState$1.compile(ModuleCompilationState.java:45)
        at io.ballerina.projects.ModuleContext.compile(ModuleContext.java:329)
        at io.ballerina.projects.PackageCompilation.compileModulesInternal(PackageCompilation.java:174)
        at io.ballerina.projects.PackageCompilation.compileModules(PackageCompilation.java:166)
        at io.ballerina.projects.PackageCompilation.from(PackageCompilation.java:95)
        at io.ballerina.projects.PackageContext.getPackageCompilation(PackageContext.java:206)
        at io.ballerina.projects.Package.getCompilation(Package.java:135)
        at io.ballerina.cli.task.CompileTask.execute(CompileTask.java:64)
        at io.ballerina.cli.TaskExecutor.executeTasks(TaskExecutor.java:40)
        at io.ballerina.cli.cmd.RunCommand.execute(RunCommand.java:163)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at io.ballerina.cli.launcher.Main.main(Main.java:58)
```

**Affected Versions:**
Ballerina Swan Lake Beta2

**OS, DB, other environment details and versions:**
MacOS BigSur

### PR DESCRIPTION
**Pull request by gimantha**
Fixes #32367
Fixes #32266

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java [hunk 1]
```diff
@@ -18161,6 +18178,13 @@ private STNode getTypeDescFromExpr(STNode expression) {
                 return expression;
             case UNARY_EXPRESSION:
                 return STNodeFactory.createSingletonTypeDescriptorNode(expression);
+            case TUPLE_TYPE_DESC:
+                STTupleTypeDescriptorNode tupleTypeDescriptorNode = (STTupleTypeDescriptorNode) expression;
+                STNode memberNodes =
+                        STNodeFactory.createNodeList(getTypeDescList((STNodeList) tupleTypeDescriptorNode
+                                .memberTypeDesc));
+                return STNodeFactory.createTupleTypeDescriptorNode(tupleTypeDescriptorNode.openBracketToken,
+                        memberNodes, tupleTypeDescriptorNode.closeBracketToken);
```
Comment (lochana-chathura) [inline compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java:18187]: What puzzles me is, why do we create a tuple-type-desc with members being expressions in the first place?

We have `getTypeDescFromExpr()` but we seem to not have `getExprFromType()`. So once tuple-type-desc node is created it is a type. 

I think we need to call `expr = getTypeDescFromExpr(expr);` before creating a TupleTypeDescriptorNode at `parseTupleTypeDescOrExprStartsWithOpenBracket()`.

In fact, I tried running unit tests with the above change and they all are parsing. @rdulmina wdyt?

Comment (rdulmina) [reply compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java:18187]: +1

File: tests/jballerina-unit-test/src/test/resources/test-src/types/tuples/tuple_negative_test.bal [hunk 2]
```diff
@@ -202,3 +202,8 @@ function testTupleToJSONAssignmentNegative() {
     jsonTest = <json[]>C;
     jsonTest = C;
 }
+
+public function testUnionsOfTupleSingletons() {
+     [1, "hello"]|[1] f; // Must not crash
+     f = [1]; // ambiguous, [1] can match for both [1, "hello"] and [1] finite tuple values;
+}
```
Comment (lochana-chathura) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/tuples/tuple_negative_test.bal:209]: Missing eof newline at the end.

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java [hunk 3]
```diff
@@ -5513,14 +5513,16 @@ private boolean isValidExprRhsStart(SyntaxKind tokenKind, SyntaxKind precedingNo
...
-            case QUESTION_MARK_TOKEN:
             case COLON_TOKEN:
             case DOT_LT_TOKEN:
             case SLASH_LT_TOKEN:
             case DOUBLE_SLASH_DOUBLE_ASTERISK_LT_TOKEN:
             case SLASH_ASTERISK_TOKEN:
             case NOT_IS_KEYWORD:
                 return true;
+            case QUESTION_MARK_TOKEN:
+                //TODO : Should fix properly #33259
```
Comment (lochana-chathura) [inline compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java:5524]: Usually, we keep a space after //, as a practice. 
```suggestion
                // TODO : Should fix properly #33259
```

**Review by lochana-chathura — COMMENTED** [review]
LGTM.

**Review by lochana-chathura — DISMISSED** [review]

**Review by lochana-chathura — APPROVED** [review]

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java [hunk 4]
```diff
@@ -15197,6 +15205,7 @@ private STNode parseTupleTypeDescOrExprStartsWithOpenBracket() {
         STNode memberEnd;
         while (!isEndOfListConstructor(peek().kind)) {
             STNode expr = parseTypeDescOrExpr();
+            expr = getTypeDescFromExpr(expr);
```
Comment (rdulmina) [inline compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java:15208]: Although we have used this, it is not clean and can cause other issues that are not in the tests eg: for samples like `Foo|[open()]` we still create tuple type with function call inside. Please create an issue to address this.

**Review by rdulmina — APPROVED** [review]

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java (hunk 4 above)
Comment (gimantha) [reply compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java:15208]: created the issue https://github.com/ballerina-platform/ballerina-lang/issues/33824

**Pull request merged**